### ISSUE TITLE
Kyma subscription and NATS infra are not in sync

### ISSUE THREAD

**Issue by raypinto** [kyma-project/kyma#14152]
**Description**

Kyma subscription and NATS infra are not in sync when we have a subscription that is not ready.

**Expected result**

Kyma subscription and NATS subscription are in sync

**Actual result**

Kyma subscription and NATS subscription are not in sync

**Steps to reproduce**

- Install kyma
- Create a subscription with 3 filters (3 subscriptions are created on NATS. You can verify it with consumers on JetStream using nats CLI)
- Edit the subscription - Delete 2 filters and have an invalid sink name
- We have only one cleanEventType but the subscription was not sync'd with the NATS infrastructure (as we still have 3 consumers instead of 1)

**Troubleshooting**

The reason for this is we do the sinkValidation before syncing the subscription with the NATS infra. We should do it regardless of an invalid sink.

**Comment by raypinto** [comment]
Although [this bug](https://github.com/kyma-project/kyma/issues/14154) was resolved, the problem of syncing the subscription with the NATS infra is still a problem and needs to be addressed as part of this issue.

If we perform the `step to reproduce` as mentioned [here](https://github.com/kyma-project/kyma/issues/14154), when we change the filter value in the spec to an invalid event Type value or invalid prefix, we see that the cleanEventTypes are empty. But the subscriptions (or consumers) on the server are still present.

**Comment by nachtmaar** [comment]
@raypinto I am a bit confused why you are talking about `cleanEventTypes` since they are defined as the `filter's event types after cleanup for use with the configured backend`, hence not important in our bug ? 
Isn't the only question whether the NATS subscriptions are created or not ? 

**From my understanding:**
- Each filter in the Subscription results in one NATS subscription
  - All NATS subscriptions share the sink from the eventing Subscription
 
**Case 1**: Should we delete all the 3 existing NATS subscriptions on the infra ? 
- Since the filters are defined on the eventing Subscription they should exist on NATS as well. However updating them with the invalid sink does not make sense I think 🤔 

**Case 2**: Should we leave the Subscription to contain the 3 valid subscriptions that are still present on the infra but not create the fourth invalid filter as a NATS subscription ? 
- I would say so. Creating a NATS subscription with an invalid sink does not make any sense 🤔

**Comment by nachtmaar** [comment]
> @k15r Q: Say we have a subscription with 3 valid filters and the respective subscriptions are created on the infra. The user adds the fourth filter to the same subscription which is invalid. Should we delete all the 3 existing subscriptions on the infra and show cleanEventTypes as empty **OR** should we leave the cleanEventTypes to contain the 3 valid subscriptions that are still present on the infra but not create the fourth invalid filter? In both cases the subscription is still not ready.

2 tests were written to check the behaviour of this. They can be found in [this file](https://github.com/nachtmaar/kyma/blob/6ef6bdeef092456307170a437d60e468125735c7/components/eventing-controller/controllers/subscription/jetstream/reconciler_test.go#L336) or just below:

<details>
  <summary>NATS subscriptions are not in sync then adding a filter with invalid event type

**TL;DR**: one invalid event-type leads to the problem that **NO clean event-type is shown** (nil list)
</summary>

```go
{
		// Ensure subscriptions on NATS are not changed then the event type is changed to an invalid format.
		// issue: https://github.com/kyma-project/kyma/issues/12979
		name: "NATS subscriptions are not in sync then adding a filter with invalid event type",
		givenSubscriptionOpts: []reconcilertesting.SubscriptionOpt{
			reconcilertesting.WithFilter(reconcilertesting.EventSource, fmt.Sprintf("%s.nonexistingapp.order.created.v1", validEventTypePrefix)),
			// valid sink
			reconcilertesting.WithSinkURLFromSvc(ens.SubscriberSvc),
		},
		wantBefore: utils.Want{
			K8sSubscription: []gomegatypes.GomegaMatcher{
				reconcilertesting.HaveSubscriptionReady(),
				// for each filter we want to have a clean event type
				reconcilertesting.HaveCleanEventTypes([]string{
					fmt.Sprintf("%s.nonexistingapp.order.created.v1", validEventTypePrefix),
				}),
			},
			K8sEvents: nil,
			// ensure that each filter results in a NATS consumer
			NatsSubscriptions: map[string][]gomegatypes.GomegaMatcher{
				fmt.Sprintf("%s.nonexistingapp.order.created.v1", validEventTypePrefix): {
					natstesting.BeExistingSubscription(),
					natstesting.BeValidSubscription(),
					natstesting.BeJetStreamSubscriptionWithSubject(fmt.Sprintf("%s.nonexistingapp.order.created.v1", validEventTypePrefix)),
				},
			},
		},
		changeSubscription: func(subscription *eventingv1alpha1.Subscription) {
			subscription.Spec.Filter.Filters = nil
			// keep the filter
			reconcilertesting.AddFilter(reconcilertesting.EventSource, fmt.Sprintf("%s.nonexistingapp.order.created.v1", validEventTypePrefix), subscription)

			// on purpose, we do use an invalid eventTypePrefix
			invalidEventTypePrefix := "sap.kyma.custom.ups"
			// add a new filter
			reconcilertesting.AddFilter(reconcilertesting.EventSource, fmt.Sprintf("%s.nonexistingapp.order.created.v1", invalidEventTypePrefix), subscription)
		},
		wantAfter: utils.Want{
			K8sSubscription: []gomegatypes.GomegaMatcher{
				reconcilertesting.HaveSubscriptionNotReady(),
				reconcilertesting.HaveCleanEventTypes([]string{
					// only the first event type can be cleaned, the second uses an invalid event type prefix
					fmt.Sprintf("%s.nonexistingapp.order.created.v1", validEventTypePrefix),
				}),
			},
			K8sEvents: nil,
			NatsSubscriptions: map[string][]gomegatypes.GomegaMatcher{
				// the first filter is still present as a NATS subscription
				fmt.Sprintf("%s.nonexistingapp.order.created.v1", validEventTypePrefix): {
					natstesting.BeExistingSubscription(),
					natstesting.BeValidSubscription(),
					natstesting.BeJetStreamSubscriptionWithSubject(fmt.Sprintf("%s.nonexistingapp.order.created.v1", validEventTypePrefix)),
				},
				// the filter with invalid event type is not created on NATS because it is using the wrong event type prefix
				"sap.kyma.custom.ups.nonexistingapp.order.created.v1": {
					gomega.Not(natstesting.BeExistingSubscription()),
				},
			},
		},
	},

```

</details> 


<details>
  <summary>NATS subscriptions are not in sync then changing the event-type to an invalid format

**TL;DR**: the **NATS subscription IS NOT deleted** if the event-type is changed to an invalid format</summary>

```go
	{
		// Ensure subscriptions on NATS are not changed when the event type is changed to an invalid format.
		// issue: https://github.com/kyma-project/kyma/issues/12979
		name: "NATS subscriptions are not in sync then changing the event-type to an invalid format",
		givenSubscriptionOpts: []reconcilertesting.SubscriptionOpt{
			reconcilertesting.WithFilter(reconcilertesting.EventSource, fmt.Sprintf("%s.nonexistingapp.order.created.v1", validEventTypePrefix)),
			// valid sink
			reconcilertesting.WithSinkURLFromSvc(ens.SubscriberSvc),
		},
		wantBefore: utils.Want{
			K8sSubscription: []gomegatypes.GomegaMatcher{
				reconcilertesting.HaveSubscriptionReady(),
				// for each filter we want to have a clean event type
				reconcilertesting.HaveCleanEventTypes([]string{
					fmt.Sprintf("%s.nonexistingapp.order.created.v1", validEventTypePrefix),
				}),
			},
			K8sEvents: nil,
			// ensure that each filter results in a NATS consumer
			NatsSubscriptions: map[string][]gomegatypes.GomegaMatcher{
				fmt.Sprintf("%s.nonexistingapp.order.created.v1", validEventTypePrefix): {
					natstesting.BeExistingSubscription(),
					natstesting.BeValidSubscription(),
					natstesting.BeJetStreamSubscriptionWithSubject(fmt.Sprintf("%s.nonexistingapp.order.created.v1", validEventTypePrefix)),
				},
			},
		},
		changeSubscription: func(subscription *eventingv1alpha1.Subscription) {
			subscription.Spec.Filter.Filters = nil
			// on purpose, we do use an invalid eventTypePrefix
			invalidEventTypePrefix := "sap.kyma.custom.ups"
			reconcilertesting.AddFilter(reconcilertesting.EventSource, fmt.Sprintf("%s.nonexistingapp.order.created.v1", invalidEventTypePrefix), subscription)
		},
		wantAfter: utils.Want{
			K8sSubscription: []gomegatypes.GomegaMatcher{
				reconcilertesting.HaveSubscriptionNotReady(),
				// the prefix is wrong, it cannot be cleaned
				reconcilertesting.HaveCleanEventTypes(nil),
			},
			K8sEvents: nil,
			// the subscription is not ready and changing the filter in the NATS subscription does not make sense because the wrong event type prefix is used.
			NatsSubscriptions: map[string][]gomegatypes.GomegaMatcher{
				// TODO: delete the NATS subscription
				fmt.Sprintf("%s.nonexistingapp.order.created.v1", validEventTypePrefix): {
					gomega.Not(natstesting.BeExistingSubscription()),
validEventTypePrefix)),
				},
			},
		},
	},
},

```

</details> 

@raypinto @k15r and me had a discussion on the topic and @k15r proposed to write a validation webhook for the Subscription CR. This way the problems which are revealed by the 2 tests can be mitigated because the webhook would decline to set an invalid event type.

**Comment by nachtmaar** [comment]
After the last team discussion about the possible options on https://github.com/kyma-project/kyma/pull/14742#discussion_r914826639, @k15r raised the concern that this might lead to **event loss**.

The problem is the following:
- If the NATS Subscription is **synced independent** of the **sink**, then this allows setting an **invalid sink** (e.g. pointing to a svc which does not exist) in the NATS Subscription.
- The Kyma Subscription will be marked as not ready by the controller, but still the sink will be set in the NATS Subscription
- That means that the dispatcher (part of the Eventing-Controller) will try sending the event to a sink which is not able to receive the event (because in this case no one is listening for events on the defined sink).
- The number of retries for NATS JetStream is set to [100](https://github.com/kyma-project/kyma/blob/317ed95ca93e83297bd4491f7e877838064b9fa8/components/eventing-controller/pkg/handlers/jetstream.go#L33)
- Sending to a not existing sink increases the retry counter and ultimately stopped as soon as the retry counter is reached.

**Comment by nachtmaar** [comment]
Blocked until https://github.com/kyma-project/kyma/issues/15087 is implemented
